crates/goose/src/agents/code_execution_extension.rs

```diff
@@ -0,0 +1,905 @@
+    }
+}
+
+pub struct CodeExecutionClient {
+    info: InitializeResult,
+    context: PlatformExtensionContext,
+}
+
+impl CodeExecutionClient {
+    pub fn new(context: PlatformExtensionContext) -> Result<Self> {
+        let info = InitializeResult {
+            protocol_version: ProtocolVersion::V_2025_03_26,
+            capabilities: ServerCapabilities {
+                tools: Some(ToolsCapability {
+                    list_changed: Some(false),
+                }),
+                resources: None,
+                prompts: None,
+                completions: None,
+                experimental: None,
+                logging: None,
+            },
+            server_info: Implementation {
+                name: EXTENSION_NAME.to_string(),
+                title: Some("Code Execution".to_string()),
+                version: "1.0.0".to_string(),
+                icons: None,
+                website_url: None,
+            },
+            instructions: Some(indoc! {r#"
+                BATCH MULTIPLE TOOL CALLS INTO ONE execute_code CALL.
+
+                This extension exists to reduce round-trips. When a task requires multiple tool calls:
+                - WRONG: Multiple execute_code calls, each with one tool
+                - RIGHT: One execute_code call with a script that calls all needed tools
+
+                Workflow:
+                    1. Use read_module("server") to discover tools and signatures
+                    2. Write ONE script that imports and calls ALL tools needed for the task
+                    3. Chain results: use output from one tool as input to the next
+            "#}.to_string()),
+        };
+
+        Ok(Self { info, context })
+    }
+
+    async fn get_tool_infos(&self) -> Vec<ToolInfo> {
+        let Some(manager) = self
+            .context
+            .extension_manager
+            .as_ref()
+            .and_then(|w| w.upgrade())
+        else {
+            return Vec::new();
+        };
+
+        match manager.get_prefixed_tools_excluding(EXTENSION_NAME).await {
+            Ok(tools) if !tools.is_empty() => {
+                tools.iter().filter_map(ToolInfo::from_mcp_tool).collect()
+            }
+            _ => Vec::new(),
+        }
+    }
+
+    async fn handle_execute_code(
+        &self,
+        arguments: Option<JsonObject>,
+    ) -> Result<Vec<Content>, String> {
+        let code = arguments
+            .as_ref()
+            .and_then(|a| a.get("code"))
+            .and_then(|v| v.as_str())
+            .ok_or("Missing required parameter: code")?
+            .to_string();
+
+        let tools = self.get_tool_infos().await;
+        let (call_tx, call_rx) = mpsc::unbounded_channel();
+        let tool_handler = tokio::spawn(Self::run_tool_handler(
+            call_rx,
+            self.context.extension_manager.clone(),
+        ));
+
+        let js_result = tokio::task::spawn_blocking(move || run_js_module(&code, &tools, call_tx))
```

Comment:
if the agent writes an infinite loop, are we toast here? for external MCP servers I think we timeout, but this is all in process so it hangs our agent I think

Reply:
We can give it a default and then let the agent try again with its own value for timeout if it needs to.